pkg/patterns/declarative/metrics.go

```diff
@@ -0,0 +1,272 @@
+
+	ManagedObjectsRecord = "managed_objects_record"
+)
+
+var (
+	reconcileCount = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Subsystem: Declarative,
+		Name:      ReconcileCount,
+		Help:      "How many times reconciliation of K8s objects managed by declarative reconciler occurs",
+	}, []string{"group_version_kind", "namespace", "name"})
+
+	reconcileFailure = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Subsystem: Declarative,
+		Name:      ReconcileFailure,
+		Help:      "How many times reconciliation failure of K8s objects managed by declarative reconciler occurs",
+	}, []string{"group_version_kind", "namespace", "name"})
+
+	managedObjectsRecord = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Subsystem: Declarative,
+		Name:      ManagedObjectsRecord,
+		Help:      "Track the number of objects in manifest",
+	}, []string{"group_version_kind", "namespace", "name"})
+)
+
+func gvkString(gvk schema.GroupVersionKind) string {
+	if len(gvk.Group) == 0 && gvk.Version == "v1" {
+		return gvk.Version + "/" + gvk.Kind
+	}
+
+	return gvk.Group + "/" + gvk.Version + "/" + gvk.Kind
+}
+
+type reconcileCounter struct {
+	groupVersionKind string
+	counterVec       *prometheus.CounterVec
+}
+
+func reconcileCounterFor(gvk schema.GroupVersionKind) reconcileCounter {
+	return reconcileCounter{groupVersionKind: gvkString(gvk), counterVec: reconcileCount}
+}
+
+func (rc reconcileCounter) reconcileWith(request reconcile.Request) {
+	rc.counterVec.WithLabelValues(rc.groupVersionKind, request.Namespace, request.Name).Inc()
+}
+
+type reconcileFailureCounter struct {
+	groupVersionKind string
+	counterVec       *prometheus.CounterVec
+}
+
+func reconcileFailureCounterFor(gvk schema.GroupVersionKind) reconcileFailureCounter {
+	return reconcileFailureCounter{groupVersionKind: gvkString(gvk), counterVec: reconcileFailure}
+}
+
+func (rfc reconcileFailureCounter) failureWith(request reconcile.Request) {
+	rfc.counterVec.WithLabelValues(rfc.groupVersionKind, request.Namespace, request.Name).Inc()
+}
+
+type objectRecorder struct {
+	groupVersionKind string
+	gaugeVec         *prometheus.GaugeVec
+}
+
+func objectRecorderFor(gvk schema.GroupVersionKind) objectRecorder {
+	return objectRecorder{groupVersionKind: gvkString(gvk), gaugeVec: managedObjectsRecord}
+}
+
+func (or objectRecorder) Set(namespace, name string, num float64) {
+	or.gaugeVec.WithLabelValues(or.groupVersionKind, namespace, name).Set(num)
+}
+
+var globalObjectTracker objectTracker = newObjectTracker()
+
+type objectTracker struct {
+	mgr        manager.Manager
+	trackedGVK map[schema.GroupVersionKind]*gvkTracker
+}
+
+// Call this before kubectl.Apply
+func (ot objectTracker) addIfNotPresent(object *manifest.Object) {
+	gvk, ns, name := object.GroupVersionKind(),
+		object.UnstructuredObject().GetNamespace(),
+		object.UnstructuredObject().GetName()
+
+	if len(ns) == 0 {
+		ns = DefaultNamespace
+	}
+
+	if gvkt, ok := ot.trackedGVK[gvk]; ok {
```

Comment:
I think you'll want to add a lock/mutex around trackedGVK. It is possible to have parallel reconciles if the user enables it in controller-runtime[1]
 
[1] https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/controller?tab=doc#Options